[14.0][FIX] Ao mapear a Linha de Operação Fiscal o método precisa considerar os casos onde o Partner do objeto não é o Partner para Faturar/Partner to Invoice

mbcosta opened the pull request:
When mapping Line Fiscal Operation the Partner used can be or not the Partner to Invoice, so to has the correct mapping the method should use the Partner to Invoice.

Ao selecionar uma Operação Fiscal nas linhas o método line_definition chamado pelo _onchange_fiscal_operation_id sempre usa o campo partner_id/res.partner do objeto https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_fiscal/models/document_fiscal_line_mixin_methods.py#L342 porém existem Casos de Uso onde o Partner selecionado no objeto pode não ser o Partner to Invoice/Parceiro para Faturar, um res.partner pode ter um Contato/res.partner definido com o campo Type como Invoice e se estiver com o campo company_type como Person o método address_get retorna esse Contato/res.partner na criação da Fatura/Invoice ou quando o objeto tem um campo especifico para isso o partner_invoice_id. Ao testar e validar esse problema também foi preciso fazer algumas correções referentes, foi preciso alterar mais de um modulo nesse PR para permitir testes mas dependendo posso ver de dividir o PR, segue abaixo os detalhes das alterações em cada modulo e os Casos de Uso mapeados, por enquanto, por isso testes e a verificação se faltou atender algum Caso de Uso são bem vindos:

l10n_br_base e l10n_br_fiscal 

* apenas inclusão de Dados de Demonstração usando dois Parceiros que não estavam sendo usados antes em outros testes um definido com um Contato de Fatura e outro Contato de Entrega

l10n_br_purchase

* esse caso pode ser remoto( qual a situação/caso de uso um pedido seria criado com o Fornecedor sendo diferente do res.partner de Faturamento? Ou usando o Endereço de Entrega? Seria o caso do dropshipping a Empresa faz o Pedido de Compra porém o Endereço de Entrega e o Cliente?) porém como ao criar uma Fatura o método do modulo purchase chama o address_get para preencher o Partner https://github.com/OCA/OCB/blob/14.0/addons/purchase/models/purchase.py#L556 isso permite que o Partner do objeto possa ser diferente do Partner to Invoice, por isso é preciso considerar o Partner retornado pelo address_get na Operação Fiscal

* inclui nos Dados de Demonstração quatro Pedidos de Compras:
   (1) Partner principal que tem um contato com Endereço de Faturamento
         Cria a Fatura com o Contato definido com o type invoice, que é diferente do Partner Principal( TODO: Nesse caso deveria considerar esse partner do Pedido o partner_shippping_id da Fatura? Isso ocorre no caso de criar a Fatura a partir de um Picking relacionado a um Pedido de Compra quando o modulo l10n_br_purchase_stock está instalado)   
   (2) Contato de um Partner definido como o Partner to Invoice
         A Fatura é criada com o mesmo Partner do Pedido
   (3) Partner que tem um Contato como Endereço de Entrega
        TODO: Deveria criar Fatura com partner_shipping_id com o Partner do Pedido de Compra?
   (4) Contato de um Partner definido como o Partner to Delivery/Parceiro de Entrega 
       TODO: Não cria a Fatura com o partner_shipping_id definido, deveria estar com o contato?

l10n_br_purchase_stock

* adicionado apenas testes, as correções necessárias para esse caso foram feitas no modulo l10n_br_stock_account por usar o address_get para definir o Partner to Invoice

l10n_br_sale

* esse caso o objeto tem um campo partner_invoice_id e quando o Partner selecionado tem um Contato Invoice o onchange preenche o campo com esse Contato mas o usuário também pode alterar o campo manualmente, por isso a Operação Fiscal precisa considerar o partner_invoice_id e não o partner_id, 

* Correção no método _prepare_invoice https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_sale/models/sale_order.py#L172 porque ao chamar o _prepare_br_fiscal_dict e fazer o update com esse dicionario acabava substituindo o Partner que é mapeado pelos métodos do modulo Sale e a Fatura acabava sendo criada com o partner_id e não o partner_invoice_id,

Apenas o modulo sale instalado

![image](https://github.com/OCA/l10n-brazil/assets/6341149/ca538b6d-12ca-4706-8aa8-113d78dce925)

![image](https://github.com/OCA/l10n-brazil/assets/6341149/3a652a5e-edd1-4730-a752-395e7512baaa)

l10n_br_sale instalado

![image](https://github.com/OCA/l10n-brazil/assets/6341149/d6190fe0-c0d1-4e42-bdfe-f7cee12e21da)

![image](https://github.com/OCA/l10n-brazil/assets/6341149/dfc5ce3f-36c9-4580-9fe5-44cb2872704c)


* adicionado nos Dados de Demontração dois Pedido de Vendas e testes para validar se os campos estão sendo preenchidos nas Faturas criadas:
   (1) Partner que possui o Endereço de Cobrança/partner_invoice_id 
   (2) Partner que possui o Endereço de Entrega/partner_shipping_id

l10n_br_sale_stock

* esse caso o Picking quando tem um Pedido de Venda/sale.order relacionado precisa usar o partner_invoice_id definido no Pedido e não o Partner do stock.picking que pode ser o partner_shipping_id do Pedido

* Correção no metodo _get_partner_to_invoice nesse mesmo sentido de usar o partner_invoice_id do Pedido

* Sobre os Casos de Uso antes desse PR ao criar uma Fatura onde o partner_id era igual ao partner_shipping_id esse era apagado e assim a Fatura era criada sem o Endereço de Entrega porém vendo que o comportamento padrão do modulo sale é de sempre criar a Fatura com o partner_shipping_id preenchido mesmo quando igual ao partner_id eu estou alterando nesse PR para fazer o mesmo ( TODO: Isso é o certo? Ou o melhor é manter como antes e alterar o l10n_br_sale para que quando o partner_id for igual ao partner_shipping_id deveria apagar o valor do campo e assim criar a Fatura sem o partner_shipping_id )

l10n_br_stock_account 

* esse Caso de Uso precisa considerar o Partner retornado pelo address_get( isso também atende o caso do l10n_br_purchase_stock ), a criação da Fatura já considera esse caso https://github.com/OCA/account-invoicing/blob/14.0/stock_picking_invoicing/models/stock_picking.py#L38 isso também resolve um TODO que existia que é a criação de uma Ordem de Entrega ou Recebimento/stock.picking sem relação com Pedido de Venda ou Compra que precisa ter um Endereço de Entrega/Recebimento diferente do Faturamento, 

* correção feita no método que retorna a Operação Fiscal padrão da empresa é preciso fazer self.env.company e não self.env.user.company_id

* Incluí nos Dados de Demonstração mais quatro pickings 
   (1) Partner principal que tem um Partner com Endereço de Entrega definido
         Caso Remoto? Um erro do usuário? Ou pode ocorrer um caso onde o usuário precisa fazer uma Entrega para um endereço diferente do que está definido com o type delivery
   (2) Partner de Endereço de Entrega
         A Fatura precisa ser criada com Partner principal
   (3) Partner principal que tem um Endereço de Faturamento definido
         A Fatura deve ser criada com o partner do addres_get e não do Picking
   (4) Partner de Endereço de Faturamento
         A Fatura criada deve ter o mesmo Partner do Picking

isso também está sendo feito para testar a criação de Faturas com Agrupamento Partner/Products nesses casos apesar das Faturas criadas acabarem sendo iguais não são agrupadas porque o método _get_picking_key considera o partner do picking https://github.com/OCA/account-invoicing/blob/14.0/stock_picking_invoicing/wizards/stock_invoice_onshipping.py#L316 TODO: é preciso avaliar se isso deve ser alterado para nesses casos Agrupar ou na localização ou mesmo no modulo stock_picking_invoicing

cc @renatonlima @rvalyi @marcelsavegnago @mileo

mbcosta committed:
[FIX] l10n_br_fiscal: proper partner_id

When mapping the Line Fiscal Operation and Taxes the Partner of the object can be or not the Partner to Invoice/Fiscal.

[FIX] l10n_br_fiscal: Document get Fiscal Partner

mbcosta committed:
[IMP] l10n_br_purchase_stock: partner_id tests

for the case of mapping Line Fiscal Operation when Partner of the object is not Partner to Invoice or Shipping.

mbcosta committed:
[FIX] l10n_br_sale: proper partner_id

Invoice should be create with partner_invoice_id not with partner_id.

[FIX] l10n_br_sale: Document get Fiscal Partner

[FIX] l10n_br_sale: Related Fiscal Partner

[FIX] l10n_br_sale: partner_invoice_id fix

When mapping the Line Fiscal Operation and Taxes the Partner of the object can be or not the Partner to Invoice, in case of Sale should it use the field partner_invoice_id.

mbcosta committed:
[FIX] l10n_br_sale_stock: proper partner_id

When Picking has a Sale Order related the Partner used to create the Invoice should be the partner_invoice_id of Sale, because the Partner of Picking can be the partner_shipping_id of Sale Order.

[FIX] l10n_br_sale_stock: Get Fiscal Partner

mbcosta committed:
[FIX] l10n_br_sale_stock: partner_invoice_id fix

When mapping the Line Fiscal Operation and Taxes the Partner of the object can be or not the Partner to Invoice, in case of Picking with related a related SO, it should use the partner_invoice_id field in Sale because the Partner of Picking can be the partner_shipping_id of the SO.

mbcosta committed:
[FIX] l10n_br_stock_account: partner_shipping_id

when creating an Invoice use the Partner mapping by the methods of module instead of Fiscal Dict and inform partner_shipping_id with Partner of Picking instead in the case of it's different of Partner to Invoice.

mileo commented:
@marcelsavegnago esse PR não tem relação com o https://github.com/OCA/l10n-brazil/pull/3185

marcelsavegnago commented:
@mileo aparentemente sim.. @mbcosta pode fazer um rebase por favor ?

mbcosta committed:
[FIX] l10n_br_purchase: proper partner_id

When mapping the Line Fiscal Operation and Taxes the Partner of the object can be or not the Partner to Invoice, in the Purchase case the code use address_get to create Invoice.

[FIX] l10n_br_purchase: Document get Fiscal Partner

mbcosta committed:
[IMP] l10n_br_purchase: demo data and tests

for the cases where the Partner of the object can be or not the Partner to Invoice.

mbcosta committed:
[IMP] l10n_br_sale: demo data and tests

for the case where the Partner of the object can be or not the Partner to Invoice.

mbcosta committed:
[FIX] l10n_br_stock_account: proper partner_id

When mapping the Line Fiscal Operation and Taxes the Partner of the object can be or not the Partner to Invoice, in the Picking case the code use address_get to create Invoice.

[FIX] l10n_br_stock_account: Get Fiscal Partner

mbcosta committed:
[IMP] l10n_br_stock_account: demo data and tests

for the case where the Partner of the object can be or not the Partner to Invoice.

marcelsavegnago commented:
> @mileo aparentemente sim.. @mbcosta pode fazer um rebase por favor ?

@mileo tem sim e com certeza é um trabalho bem mais aprofundado. @mbcosta  na PR  https://github.com/OCA/l10n-brazil/pull/3185 tem este [commit](https://github.com/OCA/l10n-brazil/pull/3185/commits/26b83e4ac12221c9fae2bf0be1b7e44b720833f4) do @rvalyi que talvez queira considerar

mbcosta commented:
valeu @mileo @marcelsavegnago , @rvalyi 

Certo Marcel posso de ver de incluir a alteração para o **related=partner_invoice_id** talvez não afete esse PR porque eu estou usando um método, aqui é preciso avaliar a melhor solução para obter o Partner Fiscal:
1) Mapeando o campo via related e um metodo
2) Apenas via metodo

Fiz o rebase porém atualizei o PR, então para quem tinha visto antes a principal diferença e que mudei o nome do metodo para _get_partner_fiscal e além do Mapeamento da Linha de Operação Fiscal também alterei dentro do Mix de Métodos Fiscais de **self.partner_id** para **self._get_partner_fiscal()**  em 

- [_compute_taxes](https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_fiscal/models/document_fiscal_line_mixin_methods.py#L193) usado no compute_taxes
- [_onchange_fiscal_operation](https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_fiscal/models/document_fiscal_line_mixin_methods.py#L365) usado pelo line_definition para obter a Linha de Operação Fiscal
- [_onchange_fiscal_operation_line_id](https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_fiscal/models/document_fiscal_line_mixin_methods.py#L377) usado no map_fiscal_tax para obter os Impostos
 
estou considerando ao invés de herdar esse método em vários objetos colocar a maioria dos casos no super **l10n_br_fiscal/models/document_fiscal_line_mixin_methods.py** :

```bash
    def _get_fiscal_partner(self):
        """
        Metodo para ser herdado e permitir mapear o Partner para os casos onde
        o objeto pode ter um Partner to Invoice diferente do Partner,
        nesses casos é preciso usar o Partner to Invoice, isso acontece por
        exemplo no Pedido de Vendas, Compras e Picking.
        """
        for record in self:
            partner = record.partner_id
            if record._name == "l10n_br_fiscal.document.line":
                partner = record.document_id.partner_id
            elif record._name == "account.move.line":
                partner = record.move_id.partner_id
            elif record._name in ("purchase.order.line", "stock.move"):
                partner = record.env["res.partner"].browse(
                    partner.address_get(["invoice"]).get("invoice")
                )
                if record.name == "stock.move":
                    if hasattr(record, "sale_line_id") and record.sale_line_id:
                        partner = record.sale_line_id.order_id.partner_invoice_id
            elif record._name == "sale.order.line":
                partner = record.order_id.partner_invoice_id

            return partner
```

Com isso ficaria desnecessário incluir nos outros módulos diminuindo código, talvez não seja o ideal mas como o PR também está com Testes depois do merge vai ser possível refatorar até ter um solução melhor, o que vocês acham?

Os Dados de Demonstração e Testes também foram separados nos Commits permitindo cherry-pick para caso alguém também queira ver de testar soluções.

marcelsavegnago commented:
@mbcosta pode fazer um rebase por favor ?

marcelsavegnago commented:
@mbcosta podemos considerar como pronto para revisão ?

mileo commented:
@mbcosta nao testei funcionalmente já testo, mas o único conceito que eu acho que pode agregar é:


No caso de um contato + empresa no partner_id, jogar no partner_invoice_id o contato da empresa. E se a empresa tiver um endereço de faturamento pegar ele.

Pq no Brasil o endereço de faturamento é meio falho e diferente do de cobrança como o @DiegoParadeda comentou no PR do Marcel.

Ou ainda adicionar uma configuração pra decidir isso.

Mas eu vou testar assim que possível junto com ele e deixo um feedback aqui.

mileo commented:
Fiz um teste no runbot e para esse caso de uso ainda não esta legal:

http://oca-l10n-brazil-14-0-pr2849-a5048d01f7a2.runboat.odoo-community.org/web#id=70&action=666&model=sale.order&view_type=form&cids=3&menu_id=491

![image](https://github.com/user-attachments/assets/8b7623c2-a59b-45a7-8559-93f5c5ed9fad)

Quando um pedido é lançado para um contato, o comercial_partner_id não é respeitado.

![image](https://github.com/user-attachments/assets/8b75d6cb-4cd7-43f7-af74-ef2d6e2e48f9)

E ai o documento fiscal fica sem dados, deve ser considerado o comercial_partner_id ou sugiro alterar o método address_get para retornar no tipo invoice o comercial partner id e poderíamos criar um parâmetro para definir isso por empresa.

mileo commented:
Na venda 50 alterei manualmente: 

![image](https://github.com/user-attachments/assets/91bf9622-e256-4dec-baaf-2c95b2fc0021)

E ai os dados do fiscal foram carregados:

![image](https://github.com/user-attachments/assets/7d8c3314-1be5-4067-aba4-e9e0a114da32)

mileo commented:
Tirando os comentários acima, parece uma implementação robusta.

DiegoParadeda commented:
@mbcosta testei aqui e tudo parece estar funcionando conforme o esperado, tá bem legal. Vejo que daria pra melhorar duas coisas só:

Primeiro seria esse ponto que o @mileo já comentou, sobre o valor default do campo partner_invoice_id na SO. No caso de ter um contado tipo **"KMEE, Diego"** e assumindo que nem Diego e nem KMEE tenham subcontatos do tipo invoice, então ao criar uma venda com o parceiro **"KMEE, Diego"** acho que seria legal que o contato **KMEE** viesse como padrão no campo partner_invoice_id. É claro que o usuário pode forçar manualmente o partner_invoice_id depois se quiser, mas imagino que o padrão seja faturar para a empresa.

O segundo ponto é que acho que a inclusão do campo partner_invoice_id no stock.picking ajudaria demais na usabilidade. Concordo 100% com os casos que você criou pra faturar um picking sem relação com Pedido de Vendas ou Compras, mas no caso eu tive que ler o código pra entender isso, acho que o usuário vai ficar bem confuso quando ver que o contato na fatura nem sempre bate com o contato do picking sabe?

De qualquer forma, já está muito legal o PR

mbcosta commented:
Pessoal vendo de responder as Revisões percebi que estava corrigindo o problema nas Linhas mas não no cabeçalho/corpo do objeto que herdam o Mixin do Documento Fiscal, exemplo era corrigido no sale.order.line mas não no sale.order, fiz a mesma solução das Linhas.

Até onde vi o campo era usado apenas no [_onchange_partner_id_fiscal](https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_fiscal/models/document_fiscal_mixin_methods.py#L96) então alterei para buscar pelo método **document_fiscal_mixin_methods.py**

De
```bash
    @api.onchange("partner_id")
    def _onchange_partner_id_fiscal(self):
        if self.partner_id:
            self.ind_final = self.partner_id.ind_final
            for line in self._get_amount_lines():
                # reload fiscal data, operation line, cfop, taxes, etc.
                line._onchange_fiscal_operation_id()
```

Para
```bash
    def _get_fiscal_partner(self):
        """
        Método para ser herdado e permitir mapear o Partner para os casos onde
        o objeto pode ter um Partner to Invoice diferente do Partner,
        nesses casos é preciso usar o Partner to Invoice, isso acontece por
        exemplo no Pedido de Vendas, Compras e Picking.
        """
        self.ensure_one()
        # Caso Fatura, o campo informado pelo usuário tem prioridade
        # e não é considerado quando o address_get retorna um Contato
        # definido como Tipo Invoice.

        return self.partner_id

    @api.onchange("partner_id")
    def _onchange_partner_id_fiscal(self):
        partner = self._get_fiscal_partner()
        if partner:
            self.ind_final = partner.ind_final
            for line in self._get_amount_lines():
                # reload fiscal data, operation line, cfop, taxes, etc.
                line._onchange_fiscal_operation_id()
```

Como tanto na Linha como no Cabeçalho o super por padrão devolve o **self.partner_id** isso deve garantir compatibilidade com todos os casos que já funcionam hoje.

Os commits estão separados para permitir cherry-picks e extrações, porém acredito que será melhor o merge com a correção tanto da Linha como do Cabeçalho juntos para evitar diferenças entre os objetos.

Os testes e Dados de Demonstração também estão em commits separados para, caso se queira, poderem ser extraídos para testar outras soluções.

Nesse PR está sendo considerado o cadastro de Endereços o Padrão do Odoo na v14, basicamente com os Endereços com o campo **is_company=False**, a questão de alterar o  o método padrão **address_get** pode ser debatido mas acho melhor em outro PR ou RFC, e talvez apenas na v16 ou superior, até para confirmar se houve alguma mudança, os testes devem ajudar nisso.

No sale.order também estava faltando alterar o related dos campos **cnpj_cpf, legal_name e ie** de **related=partner_id.cnpj_cpf** para **related=partner_invoice_id.cnpj_cpf** , em Compras/purchase.order isso é um **TODO** porque hoje não existe um campo para fazer esse related (Deveríamos criar um **fiscal_partner_id** no Mixin do Documento Fiscal? Ou teríamos alguma forma de solucionar?).

Alguns pontos a estar atento na revisão:

* É melhor separar os casos de uso dependendo do objeto, e ir validando cada caso, exemplo
       - Pedido de Vendas, todos res.partner iguais, com End. Entrega e com End de Faturamento Fiscal
       - Pedido de Compras, com End. Entrega e com End de Faturamento Fiscal
       - Ordem de Seleção Sem relação com Pedido de Vendas ou Compras
       - Ordem de Seleção Com relação com Pedido de Vendas ou Compras
       - Outros objetos ?????
  
* Para confirmar que todos estamos vendo o mesmo problema foram adicionados nos Dados de Demonstração **Cliente 2 - SP - Endereço Entrega** e **Cliente 7 - RS - Endereço Cobrança**, da forma como Odoo na v14 espera, então é melhor testar com esses cadastros para todos terem a certeza que estamos testando com a mesma configuração ( Hoje esse Dados já estão presentes, foram incluídos no PR https://github.com/OCA/l10n-brazil/pull/2868 ). 

* É preciso confirmar se todos os módulos que herdam a Linha e o Doc do MIX Fiscal estão mapeando o res.partner esperado  ( "res.partner Fiscal/Faturamento" talvez a melhor nomenclatura? ), alguns exemplos de módulos que herdam o Mixin e seria bom verificar:
   - l10n_br_sale_blanket_order https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_sale_blanket_order/models/sale_blanket_order_line.py#L69
   - l10n_br_contract https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_contract/models/contract_line.py#L31
   - etc
   
Lembrando, e para reforçar, que como o super do método _get_fiscal_partner devolve por padrão o sef.partner_id hoje não deve haver incompatibilidades.

Obrigado a todos pelas Revisões vou buscar responder logo em seguida, precisava ver de explicar os últimos commits.

mbcosta commented:
valeu @marcelsavegnago , acredito que agora sim, porque o problema passou a ser tratado tanto na Linha quanto no Cabeçalho/Corpo do objeto

mileo commented:
> Pessoal vendo de responder as Revisões percebi que estava corrigindo o problema nas Linhas mas não no cabeçalho/corpo do objeto que herdam o Mixin do Documento Fiscal, exemplo era corrigido no sale.order.line mas não no sale.order, fiz a mesma solução das Linhas.
> 
> Até onde vi o campo era usado apenas no [_onchange_partner_id_fiscal](https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_fiscal/models/document_fiscal_mixin_methods.py#L96) então alterei para buscar pelo método **document_fiscal_mixin_methods.py**
> 
> De
> 
> ```shell
>     @api.onchange("partner_id")
>     def _onchange_partner_id_fiscal(self):
>         if self.partner_id:
>             self.ind_final = self.partner_id.ind_final
>             for line in self._get_amount_lines():
>                 # reload fiscal data, operation line, cfop, taxes, etc.
>                 line._onchange_fiscal_operation_id()
> ```
> 
> Para
> 
> ```shell
>     def _get_fiscal_partner(self):
>         """
>         Método para ser herdado e permitir mapear o Partner para os casos onde
>         o objeto pode ter um Partner to Invoice diferente do Partner,
>         nesses casos é preciso usar o Partner to Invoice, isso acontece por
>         exemplo no Pedido de Vendas, Compras e Picking.
>         """
>         self.ensure_one()
>         # Caso Fatura, o campo informado pelo usuário tem prioridade
>         # e não é considerado quando o address_get retorna um Contato
>         # definido como Tipo Invoice.
> 
>         return self.partner_id
> 
>     @api.onchange("partner_id")
>     def _onchange_partner_id_fiscal(self):
>         partner = self._get_fiscal_partner()
>         if partner:
>             self.ind_final = partner.ind_final
>             for line in self._get_amount_lines():
>                 # reload fiscal data, operation line, cfop, taxes, etc.
>                 line._onchange_fiscal_operation_id()
> ```
> 
> Como tanto na Linha como no Cabeçalho o super por padrão devolve o **self.partner_id** isso deve garantir compatibilidade com todos os casos que já funcionam hoje.
> 
> Os commits estão separados para permitir cherry-picks e extrações, porém acredito que será melhor o merge com a correção tanto da Linha como do Cabeçalho juntos para evitar diferenças entre os objetos.
> 
> Os testes e Dados de Demonstração também estão em commits separados para, caso se queira, poderem ser extraídos para testar outras soluções.
> 
> Nesse PR está sendo considerado o cadastro de Endereços o Padrão do Odoo na v14, basicamente com os Endereços com o campo **is_company=False**, a questão de alterar o o método padrão **address_get** pode ser debatido mas acho melhor em outro PR ou RFC, e talvez apenas na v16 ou superior, até para confirmar se houve alguma mudança, os testes devem ajudar nisso.
> 
> No sale.order também estava faltando alterar o related dos campos **cnpj_cpf, legal_name e ie** de **related=partner_id.cnpj_cpf** para **related=partner_invoice_id.cnpj_cpf** , em Compras/purchase.order isso é um **TODO** porque hoje não existe um campo para fazer esse related (Deveríamos criar um **fiscal_partner_id** no Mixin do Documento Fiscal? Ou teríamos alguma forma de solucionar?).
> 
> Alguns pontos a estar atento na revisão:
> 
> * É melhor separar os casos de uso dependendo do objeto, e ir validando cada caso, exemplo
>   - Pedido de Vendas, todos res.partner iguais, com End. Entrega e com End de Faturamento Fiscal
>   - Pedido de Compras, com End. Entrega e com End de Faturamento Fiscal
>   - Ordem de Seleção Sem relação com Pedido de Vendas ou Compras
>   - Ordem de Seleção Com relação com Pedido de Vendas ou Compras
>   - Outros objetos ?????
> * Para confirmar que todos estamos vendo o mesmo problema foram adicionados nos Dados de Demonstração **Cliente 2 - SP - Endereço Entrega** e **Cliente 7 - RS - Endereço Cobrança**, da forma como Odoo na v14 espera, então é melhor testar com esses cadastros para todos terem a certeza que estamos testando com a mesma configuração ( Hoje esse Dados já estão presentes, foram incluídos no PR [[14.0][IMP] l10n_br_base, l10n_br_fiscal: Inclusão de um contato de Cobrança e outro de Entrega nos Dados de Demonstração #2868](https://github.com/OCA/l10n-brazil/pull/2868) ).
> * É preciso confirmar se todos os módulos que herdam a Linha e o Doc do MIX Fiscal estão mapeando o res.partner esperado  ( "res.partner Fiscal/Faturamento" talvez a melhor nomenclatura? ), alguns exemplos de módulos que herdam o Mixin e seria bom verificar:
>   
>   * l10n_br_sale_blanket_order https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_sale_blanket_order/models/sale_blanket_order_line.py#L69
>   * l10n_br_contract https://github.com/OCA/l10n-brazil/blob/14.0/l10n_br_contract/models/contract_line.py#L31
>   * etc
> 
> Lembrando, e para reforçar, que como o super do método _get_fiscal_partner devolve por padrão o sef.partner_id hoje não deve haver incompatibilidades.
> 
> Obrigado a todos pelas Revisões vou buscar responder logo em seguida, precisava ver de explicar os últimos commits.

Legal, dei uma testada hj e amanhã vou pedir para um cliente validar em dev.

Fiz esse módulo rapidinho aqui para alterar o comportamento do onchange: 

https://github.com/kmee/kmee-odoo-addons/blob/14.0/sale_enforce_commercial_partner/models/sale_order.py

posso jogar no sale-workflow em breve.

rvalyi commented:
@mileo olhando rapidamente seu módulo, eu acho que faz sentido propor no sale-workflow. (Porem me parece que na versões 16 e 17 pelo menos o proprio modulo sale ja faz isso), apenas teria que chamar o super de forma simples super() e add_list não fica uma abreviação legal, eu acho melhor botar address_list mesmo, pois add apenas sugere "adição".

WesleyOliveira98 reviewed (approved):
LGTM

mbcosta commented:
Pessoal foi feito um rebase para atualizar o PR, a única alteração feita foi no **l10n_br_sale/tests/test_l10n_br_sale.py** https://github.com/akretion/l10n-brazil/blob/14.0-FIX-mapping_line_op_fiscal_with_partner_to_invoice/l10n_br_sale/tests/test_l10n_br_sale.py#L658 para diminuir o tamanho da linha que o Ruff passou a validar, apenas para avisar quem havia testado e Revisado antes, valeu @WesleyOliveira98 pela revisão, sobre os outros pontos:

@mileo 

"No caso de um contato + empresa no partner_id, jogar no partner_invoice_id o contato da empresa. E se a empresa tiver um endereço de faturamento pegar ele."

Isso já acontece, se o Contato estiver com o **is_company=False**

"Fiz um teste no runbot e para esse caso de uso ainda não esta legal"

Para confirmar que todos estamos vendo o mesmo problema foram adicionados nos Dados de Demonstração Cliente 2 - SP - Endereço Entrega e Cliente 7 - RS - Endereço Cobrança, seria bom testar com esses dados, a forma que é feito o cadastro interfere no resultado.

Sobre o modulo no **sale-workflow** é importante perceber que pelo PR no Odoo ( parece ser referente a Bélgica ) e o comentário  "Many countries outside of EU must detect fiscal position based on delivery address(e.g.USA)" que o @rvalyi mencionou no outro PR, o caso Brasil até onde sei é diferente, porque isso não está ligado ou definido por um Mapeamento Fiscal o Usuário sempre vai precisar definir quem é o **res.partner** que deverá ser usado como **res.partner Fiscal** ou no objeto especifico ou no Cadastro do Cliente definindo como o **Invoice Address**, porque mesmo tendo esse **Invoice Address** pode ser preciso criar Documentos Fiscais para outros Contatos e isso não é por uma questão de Mapeamento Fiscal, o código não tem como identificar todos os casos.

O Odoo por algum motivo( não sei dizer qual ) a partir da v14 considera que todos os Contatos devem ser  **is_company=False** para que o método **address_get** chamado pelos onchanges e outros métodos retorne esses Contatos, uma forma de resolver seria sobre escrever o address_get para ignorar o campo **is_company** e fazer isso no modulo l10n_br_base para isso ocorrer em toda Localização, mas como escrevi antes nesse PR estou considerando a forma de cadastro que o Odoo espera, o caso Padrão, o PR já tem muitas alterações e isso pode ser feito de forma independente mas precisa ser avaliado.

@DiegoParadeda 

"Primeiro seria esse ponto que o @mileo já comentou, sobre o valor default do campo partner_invoice_id na SO."

Isso já deve ocorrer com o código atual, depende da forma que foi feito o Cadastro dos res.partner, se um Cliente/res.partner tem um Contato/res.partner definido como **Invoice Address** e o **is_company=False** ( em inglês apenas para evitar dúvida, porque na tradução ficou como Endereço de Cobrança, o que como você apontou gera dúvidas sobre a que se refere )

"mas no caso eu tive que ler o código pra entender isso, acho que o usuário vai ficar bem confuso quando ver que o contato na fatura nem sempre bate com o contato do picking sabe?"

Esse caso já existe hoje, o campo **partner_invoice_id** realmente poderia deixar isso mais evidente, mas é importante dizer que o principal motivo para incluir o campo é para permitir o caso em que a Empresa precisa gerar Documentos Fiscais para o Endereço de Entrega além do principal.

antoniospneto reviewed (commented):
@mbcosta

Ainda não finalizei minha revisão, mas gostaria de fazer uma observação:

Você não acha que o código está com excesso de comentários e "TODOs"? O excesso de ambos pode ser prejudicial, pois torna o código mais difícil de ler e manter. Comentários em excesso acabam distraindo e podem ficar desatualizados, enquanto um código bem escrito deve ser claro por si só. Já muitos "TODOs" podem indicar uma falta de organização ou prioridade, dando a impressão de um projeto inacabado.

Se isso for temporário enquanto a PR está em desenvolvimento, tudo bem, mas seria bom resolver esses pontos antes do merge.

mileo reviewed (approved):
LGTM

rvalyi commented:
pessoal, tou fazendo uns squashs e botando algumas coisas em ingles...

rvalyi commented:
@mbcosta tou até removando alguns (poucos) comentarios. Pois assim como o @antoniospneto comentou num outro PR seu dos boletos/CNAB, bom código deve ser claro por se só e as vezes demais comentários até atrapalham. Em alguns casos eu acho que o código já era claro suficiente. Ai tb qdo não se multiplica os commits demais, fica relativamente facil olhar o git blame e pular no PR ou no bug report relacionado para obter mais informações.

rvalyi reviewed (approved):
LGTM

rvalyi commented:
valeu pelo trabalho @mbcosta !

/ocabot merge minor

rvalyi commented:
/ocabot merge minor
